# Objective: Perform pagaoda analysis of single cell data of young and old fibroblasts 
# Figure xxx
# Use data generated by Seurat using xxx script: eset.RData
# Use knn.erromodles and varinfo_gene5500_mito10perc generated in PAGODA_step1.R
# I follow the tutorial of pagoda on: http://hms-dbmi.github.io/scde/pagoda.html

################################################################################################################
# packages required
library(Biobase)
library(scde)
library(RColorBrewer)
################################################################################################################
# upload data from PAGODA_step1
load("Single_cell_RNA-seq/Invivo_single_cell_RNAseq/Essentials/Expression_data_PAGODA/young_old_raw_counts_gene5500_mito10perc.Rdata")
load("/PAGODA/knn.error.models/knn.error.models.RData") 
load("/PAGODA/Varinfo/varinfo.RData") 

################################################################################################################
## Evaluate overdispersion of pre-defined gene sets - KEGG

setwd("/PAGODA/")
dir.create("KEGG_disease_aging")
setwd("./KEGG_disease_aging/")

# Identify pathways and gene sets that exhibit statistically significant excess of coordinated variability.
# Specifically, for each gene set,test whether the amount of variance explained by the first principal
# component significantly exceed the background expectation. Test both pre-defined gene sets as well as
# 'de novo' gene sets whose expression profiles are well-correlated within the given dataset.

# Upload gene sets of interest (kegg_aging.env) - KEGG_2016_disease_aging_senescence_activation
load("Single_cell_RNA-seq/Invivo_single_cell_RNAseq/Essentials/Pathways/KEGG_2016_disease_aging_activation_lower_case_for_PAGODA.RData")

# First, test pre-defined gene sets 
pwpca <- pagoda.pathway.wPCA(varinfo_gene5500_mito10perc,
                             kegg_aging.env,
                             n.components = 1,
                             n.cores = 10)

save(pwpca, file = "pwpca_top.aspects_KEGG_DISEASE_AGING_ACTIVATION_SENESCENCE_varnorm_knn.error.models.RData")

# Evaluate the statistical significance of the observed overdispersion for each KEGG gene set
pdf("pwpca_top.aspects_KEGG_DISEASE_AGING_ACTIVATION_SENESCENCE_varnorm_knn.error.models.pdf")
df <- pagoda.top.aspects(pwpca,
                         return.table = TRUE,
                         plot = TRUE,
                         z.score = 1.96)
dev.off()

# Second, test de novo gene sets
pdf("clpca_KEGG_DISEASE_AGING_ACTIVATION_SENESCENCE_novel_gene_clusters.pdf", width = 10, height = 5)
clpca <- pagoda.gene.clusters(varinfo_gene5500_mito10perc,
                              n.cores = 10,
                              plot = TRUE)
dev.off()

save(clpca, file = "clpca_KEGG_DISEASE_AGING_ACTIVATION_SENESCENCE_novel_gene_clusters.RData")

pdf("clpca_top.aspects_KEGG_DISEASE_AGING_ACTIVATION_SENESCENCE_de_novo_clusters_varnorm_knn.error.models.pdf")
df <- pagoda.top.aspects(pwpca,
                         clpca,
                         return.table = TRUE,
                         plot = TRUE,
                         z.score = 1.96)
dev.off()

################################################################################################################
# Correcting for cell cycle effect

setwd("/PAGODA/Varinfo/")

cc.pattern <- pagoda.show.pathways(c("Cell cycle", "DNA replication"), 
                                   varinfo_gene5500_mito10perc, 
                                   kegg_aging.env, 
                                   show.cell.dendrogram = FALSE, 
                                   cell.clustering = NULL, 
                                   showRowLabels = FALSE)
# subtract the pattern
varinfo_gene5500_mito10perc_disease_aging.cc <- pagoda.subtract.aspect(varinfo_gene5500_mito10perc, 
                                                         cc.pattern)

save(varinfo_gene5500_mito10perc_disease_aging.cc, file = "varinfo_gene5500_mito10perc_disease_aging.cc.RData")


################################################################################################################
################################################################################################################
## Correcting for cell cycle effect
# Evaluate overdispersion of pre-defined gene sets - KEGG correcting for Cell Cycle

setwd("/PAGODA/")
dir.create("KEGG_SENESCENCE_correctingCELLCYCLE")
setwd("./KEGG_SENESCENCE_correctingCELLCYCLE/")


# Test KEGG gene sets 
pwpca_disease_aging.cc <- pagoda.pathway.wPCA(varinfo_gene5500_mito10perc_disease_aging.cc, 
                                kegg_aging.env, 
                                n.components = 1, 
                                n.cores = 10)


save(pwpca_disease_aging.cc, file = "pwpca_top.aspects_KEGG_DISEASE_AGING_ACTIVATION_SENESCENCE_correctingCELLCYCLE_varnorm_knn.error.models.RData")

# Evaluate the statistical significance of the observed overdispersion for each KEGG gene set
pdf("pwpca.cc_top.aspects_KEGG_DISEASE_AGING_ACTIVATION_SENESCENCE_correctingCELLCYCLE_varnorm_knn.error.models.pdf")
df <- pagoda.top.aspects(pwpca_disease_aging.cc, 
                         return.table = TRUE, 
                         plot = TRUE, 
                         z.score = 1.96)
dev.off()

# Second, test de novo gene sets 
pdf("clpca_KEGG_DISEASE_AGING_ACTIVATION_SENESCENCE_correctingCELLCYCLE_novel_gene_clusters.pdf", width = 10, height = 5)
clpca_disease_aging.cc <- pagoda.gene.clusters(varinfo_gene5500_mito10perc_disease_aging.cc,
                              n.cores = 10,
                              plot = TRUE)
dev.off()

save(clpca_disease_aging.cc, file = "clpca_KEGG_DISEASE_AGING_ACTIVATION_SENESCENCE_correctingCELLCYCLE_novel_gene_clusters.RData")

pdf("clpca_top.aspects_KEGG_DISEASE_AGING_ACTIVATION_SENESCENCE_de_novo_clusters_correctingCELLCYCLE_varnorm_knn.error.models.pdf")
df <- pagoda.top.aspects(pwpca_disease_aging.cc, 
                         clpca_disease_aging.cc, 
                         return.table = TRUE, 
                         plot = TRUE, 
                         z.score = 1.96)
dev.off()
